File: grpcclient/grpcClient.go
package grpcclient

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"

	"google.golang.org/grpc"

	"musicas-client/proto"
)

var testCounts = []int{100, 200, 300}

func testLoad(client interface{}, method string, args interface{}, name string, numTest int) {
	start := time.Now()

	var wg sync.WaitGroup
	for i := 0; i < numTest; i++ {
		wg.Add(1)
		func() {
			defer wg.Done()
			switch method {
			case "LerUsuarios":
				c := client.(proto.UsuarioServiceClient)
				_, err := c.LerUsuarios(context.Background(), args.(*proto.Empty))
				if err != nil {
					log.Fatalf("Error: %v", err)
				}
			case "LerMusicas":
				c := client.(proto.MusicaServiceClient)
				_, err := c.LerMusicas(context.Background(), args.(*proto.Empty))
				if err != nil {
					log.Fatalf("Error: %v", err)
				}
			case "LerPlaylists":
				c := client.(proto.PlaylistServiceClient)
				_, err := c.LerPlaylists(context.Background(), args.(*proto.Empty))
				if err != nil {
					log.Fatalf("Error: %v", err)
				}
			default:
				log.Fatalf("Unknown method: %v", method)
			}
		}()
	}
	wg.Wait()

	end := time.Now()
	fmt.Printf("%s load test with %d requests completed in %v seconds\n", name, numTest, end.Sub(start).Seconds())
}

func RunTests() {
	conn, err := grpc.Dial("localhost:5003", grpc.WithInsecure())
	if err != nil {
		log.Fatalf("Failed to connect: %v", err)
	}
	defer conn.Close()

	clientUsuario := proto.NewUsuarioServiceClient(conn)
	clientMusica := proto.NewMusicaServiceClient(conn)
	clientPlaylist := proto.NewPlaylistServiceClient(conn)

	for _, numTest := range testCounts {
		fmt.Printf("Running tests with %d requests:\n", numTest)

		testLoad(clientUsuario, "LerUsuarios", &proto.Empty{}, "Ler Usuários", numTest)
		testLoad(clientMusica, "LerMusicas", &proto.Empty{}, "Ler Músicas", numTest)
		testLoad(clientPlaylist, "LerPlaylists", &proto.Empty{}, "Listar Playlists", numTest)
	}
}

File: python/clients/grpc_client.py
import grpc
import time

from clients.proto import dtos_pb2, dtos_pb2_grpc

TEST_COUNTS = [100, 200, 300]


def test_load(method, name, num_test):
    start = time.time()

    for i in range(num_test):
        try:
            response = method(dtos_pb2.Empty())  # type: ignore
        except grpc.RpcError as e:
            print(f"Request {i} failed: {e}")

    end = time.time()
    print(
        f"{name} load test with {num_test} requests completed in {end - start:.6f} seconds"
    )


def run_tests():
    channel = grpc.insecure_channel("localhost:5003")
    usuario_stub = dtos_pb2_grpc.UsuarioServiceStub(channel)
    musica_stub = dtos_pb2_grpc.MusicaServiceStub(channel)
    playlist_stub = dtos_pb2_grpc.PlaylistServiceStub(channel)

    for num_test in TEST_COUNTS:
        print(f"Running tests with {num_test} requests:")
        test_load(usuario_stub.LerUsuarios, "Usuarios", num_test)
        test_load(musica_stub.LerMusicas, "Musicas", num_test)
        test_load(playlist_stub.LerPlaylists, "Playlists", num_test)


if __name__ == "__main__":
    run_tests()
